Codechange: [Script] Don't expose static buffers outside of ScriptText

glx22 opened the pull request:
Supersede #10479.

## Motivation / Problem
As shown in #10479, in `ScriptLeagueTable::New()` encoded_title directly store the static buffer without copying its value first, so in the end encoded_title value will be the result of the last `GetEncodedText()` call (footer in the example given).
<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description
Convert the static buffer content into an `std::string` on return.
~~The return type can't be `std::string` directly due to the special handling of SCRIPT_TEXT_MAX_PARAMETERS (which in theory can't never happen because it's already checked when adding the parameters), so I kept the current mechanism using `std::optional<std::string>`.
I also extracted the "too many parameters" check from `EnforcePreconditionEncodedText()`, to allow checking the validity when an empty string is allowed (again not really needed but safer).~~ Removed `ScriptError::ERR_PRECONDITION_TOO_MANY_PARAMETERS` and replaced its handling with a fatal error, as it's really not supposed to happen.
Also used `std::string` to store RawString value, removing the manual stredup/free.
Updated the docs to reflect empty string being allowed when nullptr is.
<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

frosch123 commented on src/script/api/script_text.cpp:
```diff
@@ -17,15 +17,10 @@
 #include "../../safeguards.h"
 
-RawText::RawText(const char *text) : text(stredup(text))
+RawText::RawText(const char *text) : text(text)
```
Can `text` be `nullptr`?

glx22 replied:
Based of my understanding of https://github.com/OpenTTD/OpenTTD/blob/master/cmake/scripts/SquirrelExport.cmake#L38-L47 it cannot be `nullptr`

frosch123 commented on src/script/api/script_text.hpp:
```diff
@@ -21,17 +23,17 @@ class Text : public ScriptObject {
 	 * @api -all
 	 */
-	virtual const char *GetEncodedText() = 0;
+	virtual std::optional<const std::string> GetEncodedText() = 0;
```
The semantics of `optional<const T>` confuses me. The result is a value, no reference, so `const` is weird.

glx22 replied:
If we replace the special handling of SCRIPT_TEXT_MAX_PARAMETERS by some kind of assertion check, it's not even needed to use std::optional. The number of parameters is already checked in constructor (throws "invalid parameter"), and AddParam() or SetParam() can't add more than the limit (no throw, the call just fails and says "unknown error").

Maybe I should just throw in AddParam() and SetParam() to give useful info, and also find a way to kill the script in case it managed to add more than SCRIPT_TEXT_MAX_PARAMETERS (the special handling requiring std::optional), or just assume it can't happen, but I don't like removing a safety check.

Edit: I replaced handling of SCRIPT_TEXT_MAX_PARAMETERS as a precondition error by a fatal error, as it makes more sense that way. This removes the need to use std::optional.

glx22 committed:
Change: [Script] A ScriptText with too many parameters is now a fatal error

It should never happen as adding/setting parameters already checks that anyway.

glx22 committed:
Codechange: [Script] Don't expose static buffers outside of ScriptText
